satellite/satellitedb/overlaycache.go

```diff
@@ -33,112 +33,167 @@ type overlaycache struct {
+	if newNodeCount == 0 && reputableNodeCount == 0 {
+		return nil, nil
+	}
+	safeNewNodeQuery := ""
+	newNodeArgs := []interface{}{}
+	if newNodeCount > 0 {
+		safeNewNodeQuery, newNodeArgs, err = buildConditions(ctx, criteria, true)
 		if err != nil {
-			return nil, Error.New("invalid node selection criteria version: %v", err)
+			return nil, err
 		}
-		safeQuery += `
-			AND (major > ? OR (major = ? AND (minor > ? OR (minor = ? AND patch >= ?))))
-			AND release`
-		args = append(args, v.Major, v.Major, v.Minor, v.Minor, v.Patch)
+	}
+	safeReputableNodeQuery, reputableNodeArgs, err := buildConditions(ctx, criteria, false)
+	if err != nil {
+		return nil, err
 	}
 
 	if !criteria.DistinctIP {
-		nodes, err = cache.queryNodes(ctx, criteria.ExcludedIDs, count, safeQuery, args...)
+		newNodeQuery, moreNewNodeArgs := buildSelection(ctx, newNodeCount, safeNewNodeQuery)
+		newNodeArgs = append(newNodeArgs, moreNewNodeArgs...)
+
+		reputableNodeQuery, moreReputableNodeArgs := buildSelection(ctx, reputableNodeCount, safeReputableNodeQuery)
+		reputableNodeArgs = append(reputableNodeArgs, moreReputableNodeArgs...)
+
+		finalQuery := newNodeQuery + " UNION ALL " + reputableNodeQuery
+		finalArgs := []interface{}{}
+		finalArgs = append(finalArgs, newNodeArgs...)
+		finalArgs = append(finalArgs, reputableNodeArgs...)
+
+		rows, err := cache.db.Query(ctx, cache.db.Rebind(finalQuery), finalArgs...)
 		if err != nil {
-			return nil, err
+			return nil, Error.Wrap(err)
+		}
+		defer func() { err = errs.Combine(err, rows.Close()) }()
+
+		var nodes []*overlay.SelectedNode
+		for rows.Next() {
+			var node overlay.SelectedNode
+			node.Address = &pb.NodeAddress{Transport: pb.NodeTransport_TCP_TLS_GRPC}
+			err = rows.Scan(&node.LastNet, &node.ID, &node.Address.Address, &node.LastIPPort)
+			if err != nil {
+				return nil, err
+			}
+			nodes = append(nodes, &node)
 		}
 		return nodes, nil
 	}
 
+	totalCount := newNodeCount + reputableNodeCount
+	receivedNewNodeCount := 0
+	receivedNodeNetworks := make(map[string]struct{})
+
 	for i := 0; i < 3; i++ {
-		moreNodes, err := cache.queryNodesDistinct(ctx, criteria.ExcludedIDs, criteria.ExcludedNetworks, count-len(nodes), safeQuery, criteria.DistinctIP, args...)
+		finalQuery := ""
+		finalArgs := []interface{}{}
+
+		// temporary arg slices are needed for each iteration, otherwise we'll end up with too many args appended
+		tempNewNodeArgs := []interface{}{}
+		tempNewNodeArgs = append(tempNewNodeArgs, newNodeArgs...)
+		tempReputableNodeArgs := []interface{}{}
+		tempReputableNodeArgs = append(tempReputableNodeArgs, reputableNodeArgs...)
+
+		if receivedNewNodeCount < newNodeCount {
+			newNodeQuery, moreNewNodeArgs := buildSelectionDistinct(ctx, criteria.ExcludedNetworks, newNodeCount, safeNewNodeQuery, true)
+			tempNewNodeArgs = append(tempNewNodeArgs, moreNewNodeArgs...)
+			finalArgs = append(finalArgs, tempNewNodeArgs...)
+			finalQuery = newNodeQuery + "  UNION ALL "
```

Comment:
Why does it need to do a union here?

Reply:
the union all is so that the new node query string can be concatenated to the reputable node query string below. if you would prefer, I could move the union all to line 109.

Reply:
At this part I don't think it should need to do a union any more since the initial node selection failed so it should fallback to a wider implementation.

Reply:
I'm not sure what you mean by "since the initial node selection failed" or "it should fallback to a wider implementation" - could you clarify?

Reply:
The initial selection tries to get the necessary ratio of new vs old. However when that fails it should try to select either to fill the necessary limit.

Reply:
I just realized that it's also in the old implementation and I cannot quite understand why.

Reply:
Found the reason https://github.com/storj/storj/pull/1890/files#r284444400